domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+VALUES ($M.block_device_uuid, $M.machine_uuid)
+`
+	insertJoinStmt, err := sqlair.Prepare(insertJoinQuery, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for uuid := range blockDevicesByUUID {
+		if err := tx.Query(ctx, insertJoinStmt, sqlair.M{
+			"machine_uuid":      machineUUID,
+			"block_device_uuid": uuid.String(),
+		}).Run(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func blockDevicesChanged(oldDevices, newDevices []blockdevice.BlockDevice) bool {
+	if len(oldDevices) != len(newDevices) {
+		return true
+	}
+	for _, o := range oldDevices {
+		var found bool
+		for _, n := range newDevices {
+			if reflect.DeepEqual(o, n) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return true
+		}
+	}
+	return false
+}
+
+// RemoveMachineBlockDevices removes all the block devices for the specified machine.
+// It is the same as calling SetMachineBlockDevices with an empty list, but does not
+// error if the machine life is Dead.
+func (st *State) RemoveMachineBlockDevices(ctx context.Context, machineId string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		q := `
+SELECT machine.uuid AS &M.machine_uuid
+FROM   machine
+WHERE  machine.machine_id = $M.machine_id
+`
+		stmt, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		result := sqlair.M{}
+		err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(err)
+		}
+		if len(result) == 0 {
+			return fmt.Errorf("machine %q %w", machineId, errors.NotFound)
+		}
+		machineUUID := result["machine_uuid"].(string)
+		if err := removeMachineBlockDevices(ctx, tx, machineUUID); err != nil {
+			return errors.Annotatef(err, "removing block devices on machine %q (%s)", machineId, machineUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func removeMachineBlockDevices(ctx context.Context, tx *sqlair.TX, machineUUID string) error {
+	machineUUIDParam := sqlair.M{"machine_uuid": machineUUID}
+
+	// TODO(wallyworld) - sqlair doesn't support IN clauses yet.
+	// Ideally, we'd first get the block device UUIDs to delete
+	// and use IN clauses with the delete queries
+	// This avoids the need for a potentially inefficient
+	// select distinct over a (maybe) large number of rows when
+	// deleting from block_device. In practice, it would be at
+	// most no more than up to 1000 rows in the extreme case.
+
+	linkDeleteQuery := fmt.Sprintf(`
+DELETE 
+FROM  block_device_link_device
+WHERE block_device_uuid IN (
+    SELECT DISTINCT block_device_uuid
+    FROM            block_device_machine bdm
+    WHERE           bdm.machine_uuid = $M.machine_uuid
+)`)
+
+	deleteStmt, err := sqlair.Prepare(linkDeleteQuery, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
```

Comment:
I think we can do better here. Can we annotate these errors with the operation we were trying to perform with the context id's we were trying to use?

Reply:
Not all errors need annotation. An error here is a programming error.